core/components/menu/menu.js

```diff
@@ -0,0 +1,543 @@
+
+  this.attachEvents_();
+};
+
+/**
+ * Cleans up the container before its DOM is removed from the document, and
+ * removes event handlers.  Overrides {@link Blockly.Component#exitDocument}.
+ * @override
+ */
+Blockly.Menu.prototype.exitDocument = function() {
+  // {@link #setHighlightedIndex} has to be called before
+  // {@link Blockly.Component#exitDocument}, otherwise it has no effect.
+  this.setHighlightedIndex(-1);
+
+  Blockly.Menu.superClass_.exitDocument.call(this);
+};
+
+/** @override */
+Blockly.Menu.prototype.disposeInternal = function() {
+  Blockly.Menu.superClass_.disposeInternal.call(this);
+
+  this.detachEvents_();
+};
+
+/**
+ * Adds the event listeners to the menu.
+ * @private
+ */
+Blockly.Menu.prototype.attachEvents_ = function() {
+  var el = /** @type {!EventTarget} */ (this.getElement());
+
+  this.mouseOverHandler_ = Blockly.bindEventWithChecks_(el,
+      'mouseover', this, this.handleMouseOver_, true);
+  this.mouseUpHandler_ = Blockly.bindEventWithChecks_(el,
+      'mouseup', this, this.handleMouseUp_, true);
+  this.mouseEnterHandler_ = Blockly.bindEventWithChecks_(el,
+      'mouseenter', this, this.handleMouseEnter_, true);
+  this.mouseLeaveHandler_ = Blockly.bindEventWithChecks_(el,
+      'mouseleave', this, this.handleMouseLeave_, true);
+
+  this.onKeyDownWrapper_ = Blockly.bindEventWithChecks_(el,
+      'keydown', this, this.handleKeyEvent);
+};
+
+/**
+ * Removes the event listeners from the menu.
+ * @private
+ */
+Blockly.Menu.prototype.detachEvents_ = function() {
+  Blockly.unbindEvent_(this.mouseOverHandler_);
+  Blockly.unbindEvent_(this.mouseUpHandler_);
+  Blockly.unbindEvent_(this.mouseEnterHandler_);
+  Blockly.unbindEvent_(this.mouseLeaveHandler_);
+  Blockly.unbindEvent_(this.onKeyDownWrapper_);
+};
+
+// Child component management.
+
+/**
+ * Map of DOM IDs to child controls.  Each key is the DOM ID of a child
+ * control's root element; each value is a reference to the child control
+ * itself.  Used for looking up the child control corresponding to a DOM
+ * node in O(1) time.
+ * @type {?Object}
+ * @private
+ */
+Blockly.Menu.prototype.childElementIdMap_ = null;
+
+/**
+ * Creates a DOM ID for the child control and registers it to an internal
+ * hash table to be able to find it fast by id.
+ * @param {Blockly.Component} child The child control. Its root element has
+ *     to be created yet.
+ * @private
+ */
+Blockly.Menu.prototype.registerChildId_ = function(child) {
+  // Map the DOM ID of the control's root element to the control itself.
+  var childElem = child.getElement();
+
+  // If the control's root element doesn't have a DOM ID assign one.
+  var id = childElem.id || (childElem.id = child.getId());
+
+  // Lazily create the child element ID map on first use.
+  if (!this.childElementIdMap_) {
+    this.childElementIdMap_ = {};
+  }
+  this.childElementIdMap_[id] = child;
+};
+
+/**
+ * Returns the child menu item that owns the given DOM node, or null if no such
```

Comment:
Is there a difference between a "menu item" and a "control"?

Reply:
No, I'll update all references to control with menu item, since that's the only use case here.